### PR TITLE
Fix transformers_modeling_backend for pretrained dense models

### PR DESCRIPTION
**Pull request by HosseinKaviani-H**
Fix for #3775 

***Problem***

Loading a real pretrained dense model (e.g. Qwen3-0.6B) into the HF backend produced near-random loss, and on recent PyTorch the backend crashed before training. The HF safetensors load correctly (verified byte-for-byte): the bugs are in how the backend builds/initializes the model, not in loading.

With the same Qwen3-0.6B checkpoint + data, step-1 loss was 11.08 (≈ ln(vocab), i.e. random) vs native TorchTitan qwen3's 3.63.

***Root causes fixed***

  1. RoPE inv_freq left uninitialized after meta-device init. HF rotary modules compute inv_freq in __init__; meta init + to_empty()
  zeros it and init_weights only re-inits parameters, silently disabling RoPE. Recompute it from each rotary module's rope_init_fn in
  init_states.
  2. TitanDenseModelConfig overrode the HF config. Non-None defaults (dim=4096, n_layers=32, n_heads=32, norm_eps=1e-5,
  rope_theta=10000) were injected over AutoConfig values, forcing the wrong architecture (e.g. rope_theta 1e6 → 10000 for Qwen3).
  Default these to None; set explicitly only to intentionally override (e.g. debugmodel).
  3. Tied embeddings crashed FSDP2. tok_embeddings/lm_head share a parameter under tie_word_embeddings but were in separate fully_shard
  groups (forbidden by recent FSDP2). Group tied tok_embeddings/norm/lm_head into one unit.
  4. head_dim/intermediate_size derived instead of read from the HF config. update_from_config set head_dim = hidden_size // num_heads,
  overriding Qwen3's explicit head_dim=128 with 64 and mismatching checkpoint shapes. Only derive when dim is explicitly overridden;
  otherwise keep the HF config's values.
  5. No state-dict adapter. model_registry set state_dict_adapter=None, so HF weight names were never mapped and initial_load_in_hf
  loaded nothing. Add HFTransformerStateDictAdapter (model. prefix + tied-embedding handling).

***Verification***

Loading Qwen3-0.6B into the full config, same checkpoint/data/seed as native (--debug.seed 42 --debug.deterministic):

  - Forward parity (step 1, on loaded weights): native 3.631 vs HF 3.623 — identical within bf16 tolerance (was 11.08 before).
  - Training parity (matched LR 3e-4): the curves track step-for-step with no instability:

  | step | native | HF   |
  |------|--------|------|
  | 1    | 3.63   | 3.62 |
  | 2    | 4.06   | 4.14 |
  | 5    | 4.08   | 4.28 |

  - (Residual drift is expected — two different model implementations diverge as updates accumulate.)
  - The backend previously crashed at the FSDP tied-param check on any tied model (incl. debugmodel); it now trains.

### CODE REVIEW

**Comment by acisseJZhong** [comment]
> Forward parity (step 1, on loaded weights): native 3.631 vs HF 3.623 — identical within bf16 tolerance (was 11.08 before).

can you also compare the kl divergence between titan logits and HF logits after the first step? previously @shuhuayu compared titan model vs HF model and the logits difference is within 1e-7.

**Comment by shuhuayu** [comment]
> > Forward parity (step 1, on loaded weights): native 3.631 vs HF 3.623 — identical within bf16 tolerance (was 11.08 before).
> 
> can you also compare the kl divergence between titan logits and HF logits after the first step? previously @shuhuayu compared titan model vs HF model and the logits difference is within 1e-7.

yeah, you can try a numerical test similar to `scripts/checkpoint_conversion/numerical_tests_qwen3_5.py` to check the  logits divergence to have more confidence in the numerics.

File: torchtitan/experiments/transformers_modeling_backend/__init__.py [hunk 1]
```diff
@@ -18,17 +19,25 @@
...
+    whose config differs (e.g. rope_theta=1e6 for Qwen3). Set a field explicitly
+    only to intentionally override the HF config (e.g. debugmodel sizes).
+    """
+
+    dim: int | None = None
+    n_layers: int | None = None
+    n_heads: int | None = None
     n_kv_heads: int | None = None
     vocab_size: int | None = None
     multiple_of: int = 256
```
Comment (acisseJZhong) [inline torchtitan/experiments/transformers_modeling_backend/__init__.py]: curious why do we want to intentionally override HF for this?

File: torchtitan/experiments/transformers_modeling_backend/model.py [hunk 2]
```diff
@@ -259,16 +259,25 @@ def update_from_config(
...
-                ffn_hidden_size = 4 * self.dim
-                ffn_hidden_size = int(2 * ffn_hidden_size / 3)
-                if self.ffn_dim_multiplier is not None:
-                    ffn_hidden_size = int(self.ffn_dim_multiplier * ffn_hidden_size)
-                self.intermediate_size = self.multiple_of * (
-                    (ffn_hidden_size + self.multiple_of - 1) // self.multiple_of
-                )
-
-            self.head_dim = self.dim // self.num_attention_heads
+            # When dim is explicitly overridden (e.g. debugmodel), derive the
```
Comment (acisseJZhong) [inline torchtitan/experiments/transformers_modeling_backend/model.py:268]: dim is no longer explicitly overridden now?

File: torchtitan/experiments/transformers_modeling_backend/__init__.py (hunk 1 above)
Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/__init__.py]: Do you mean for the multiple_of? That's not a HF config field - so it doesn't override anything from the AutoConfig.

**Comment by HosseinKaviani-H** [comment]
@acisseJZhong @shuhuayu  Ran a logit-divergence test (let me know if you want me to push the script as well) -backend vs plain HF from_pretrained on Qwen3-0.6B:

  - Full sequence: KL ≈ 9e-7, cosine = 1.0, top-1 = 100%
  - Last token: KL ≈ 9e-8, max_diff ≈ 2.6e-5

  So the backend logits match HF within ~1e-7, in line with @shuhuayu's earlier result.

**Comment by shuhuayu** [comment]
> @acisseJZhong @shuhuayu Ran a logit-divergence test (let me know if you want me to push the script as well) -backend vs plain HF from_pretrained on Qwen3-0.6B:
> 
> * Full sequence: KL ≈ 9e-7, cosine = 1.0, top-1 = 100%
> * Last token: KL ≈ 9e-8, max_diff ≈ 2.6e-5
> 
> So the backend logits match HF within ~1e-7, in line with @shuhuayu's earlier result.

thanks. i think the script will be very similar to the existing example so it's fine to not include it.

File: torchtitan/experiments/transformers_modeling_backend/parallelize.py [hunk 3]
```diff
@@ -317,8 +334,9 @@ def apply_fsdp(
         )
 
     # As an optimization, do not reshard_after_forward the last layers by default
-    # since FSDP would prefetch them immediately after the forward pass
-    if model.norm is not None and model.lm_head is not None:
+    # since FSDP would prefetch them immediately after the forward pass. When
+    # weights are tied, norm/lm_head are already grouped with tok_embeddings above.
+    if not weights_tied and model.norm is not None and model.lm_head is not None:
```
Comment (acisseJZhong) [inline torchtitan/experiments/transformers_modeling_backend/parallelize.py]: you've already checked model.lm_haed not None in weights_tied?

File: torchtitan/experiments/transformers_modeling_backend/model.py (hunk 2 above)
Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/model.py:268]: dim is still overridden by the debugmodel flavor (dim=256, n_heads=16, …), therefore that hits the derive path. For loaded models (full, dim=None) we skip it and keep the HF config's head_dim (Qwen3 uses 128, not hidden/heads=64). It's only set inside the if/elif, no unconditional override.

File: torchtitan/experiments/transformers_modeling_backend/parallelize.py (hunk 3 above)
Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/parallelize.py]: Not redundant - weights_tied only implies lm_head exists when it's True. In the not weights_tied branch lm_head can still be None (no .weight or PP case), so the check is still needed here. It's also the pre existing guard.

File: torchtitan/experiments/transformers_modeling_backend/__init__.py (hunk 1 above)
Comment (acisseJZhong) [reply torchtitan/experiments/transformers_modeling_backend/__init__.py]: I see worth adding a comment: to clarify why some fields have defaults.

File: torchtitan/experiments/transformers_modeling_backend/parallelize.py [hunk 4]
```diff
@@ -273,7 +273,24 @@ def apply_fsdp(
         reshard_after_forward_policy, pp_enabled
     )
 
-    if model.tok_embeddings is not None:
+    # When input/output embeddings are tied (e.g. Qwen3), tok_embeddings and
+    # lm_head share one parameter. FSDP2 forbids a parameter being managed by
+    # two FSDP groups, so they must be grouped into a single unit.
+    tok_emb_weight = getattr(model.tok_embeddings, "weight", None)
```
Comment (acisseJZhong) [inline torchtitan/experiments/transformers_modeling_backend/parallelize.py:279]: nit: anyway you can re-use https://github.com/pytorch/torchtitan/blob/4054e01c483809a017dce7ead0e464cd492bd023/torchtitan/models/common/decoder.py#L84 or add similar fields to HF Config?

Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/parallelize.py:279]: tie_word_embeddings is the HF-side flag (the adapter uses it) - enable_weight_tying is on Decoder.Config which this Config doesn't inherit. I used parameter identity on purpose - it's exactly what FSDP checks and it auto-skips PP nn.Identity placeholders (no .weight) that a flag would wrongly group. I can switch to tie_word_embeddings if you prefer (prob with an identity guard)

File: torchtitan/experiments/transformers_modeling_backend/__init__.py (hunk 1 above)
Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/__init__.py]: Done

File: torchtitan/experiments/transformers_modeling_backend/parallelize.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/experiments/transformers_modeling_backend/parallelize.py:279]: not worth it if it's getting too complicated, but at least we could align the naming?

Comment (HosseinKaviani-H) [reply torchtitan/experiments/transformers_modeling_backend/parallelize.py:279]: Done.

**Review by acisseJZhong — APPROVED** [review]
LGTM, please fix the linter before landing and make sure failing tests are not caused by this PR

**Comment by HosseinKaviani-H** [comment]
> LGTM, please fix the linter before landing and make sure failing tests are not caused by this PR

@acisseJZhong The lint failure is Pyrefly, not this PR:
1- Pyrefly excludes torchtitan/experiments/ (all my changes are there), so my files aren't even checked.
2- The 5 errors are in core files I didn't touch (distributed/utils.py, token_dispatcher.py), caused by the CI env missing its  ../pytorch search path (WARN .../pytorch does not exist) therefore can't resolve batch_invariant_ops/torchcomms/all_to_all_single types.

File: torchtitan/experiments/transformers_modeling_backend/model.py [hunk 5]
```diff
@@ -204,6 +204,11 @@ def setter(self, value):
                 self._tt_to_hf_attribute_map.update(_TT_TO_HF_MAPPINGS["moe"])
 
             for titan_name, hf_name in self._tt_to_hf_attribute_map.items():
+                # A self-mapped name (e.g. vocab_size -> vocab_size) is already
+                # stored under the correct name; creating an alias property would
+                # make its setter call setattr() on itself and recurse forever.
+                if titan_name == hf_name:
```
Comment (HosseinKaviani-H) [inline torchtitan/experiments/transformers_modeling_backend/model.py:210]: @acisseJZhong Fixed the RecursionError in the integration test: the vocab_size -> vocab_size self-mapping built a self-referential alias property that recursed on first set. Now skip alias creation when TT and HF names match. Other CI failures (pyrefly, ROCm loss-parity) are pre-existing/environmental.
